lib/recursive-methods.rb

```diff
@@ -1,49 +1,105 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the value of n
+# Space complexity: O(n) where n is the value of n
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n <= 1
+    return 1
+  else 
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s 
+  else 
+    return s[s.length - 1] + reverse(s[0..-2])
```

Comment:
because `s[0..-2]`  generates a **new** array and copies over all the elements this  solution has O(n<sup>2</sup>) time & space complexity.